src/test/rgw/rgw_multi/tests.py

```diff
@@ -1664,6 +1673,118 @@ def make_test_bucket():
     test_bilog = bilog_list(zone.zone, test_bucket.name)
     assert(len(test_bilog) > 0)
 
+def test_bucket_log_trim_after_delete_bucket():
+    zonegroup = realm.master_zonegroup()
+    zonegroup_conns = ZonegroupConns(zonegroup)
+
+    primary = zonegroup_conns.rw_zones[0]
+    secondary = zonegroup_conns.rw_zones[1]
+
+    # create a test bucket, upload some objects, and wait for sync
+    def make_test_bucket():
+        name = gen_bucket_name()
+        log.info('create bucket zone=%s name=%s', primary.name, name)
+        bucket = primary.conn.create_bucket(name)
+        for objname in ('a', 'b', 'c', 'd'):
+            k = new_key(primary, name, objname)
+            k.set_contents_from_string('foo')
+        zonegroup_meta_checkpoint(zonegroup)
+        zonegroup_bucket_checkpoint(zonegroup_conns, name)
+        return bucket
+
+    # create a 'test' bucket
+    test_bucket = make_test_bucket()
+
+    # Resharding the bucket
+    secondary.zone.cluster.admin(['bucket', 'reshard',
+        '--bucket', test_bucket.name,
+        '--num-shards', '13',
+        '--yes-i-really-mean-it'])
+
+    # Delete the objects
+    for obj in ('a', 'b', 'c', 'd'):
+        cmd = ['object', 'rm'] + primary.zone.zone_args()
+        cmd += ['--bucket', test_bucket.name]
+        cmd += ['--object', obj]
+        primary.zone.cluster.admin(cmd)
+
+
+    # delete bucket and test bilog autotrim
+    primary.conn.delete_bucket(test_bucket.name)
+
+    zonegroup_data_checkpoint(zonegroup_conns)
+    zonegroup_meta_checkpoint(zonegroup)
+
+    # run bilog trim twice to clean up previous gen
+    bilog_autotrim(secondary.zone)
+    time.sleep(config.checkpoint_delay)
+
+    # trim bilog on primary
+    bilog_autotrim(primary.zone)
+    time.sleep(config.checkpoint_delay)
+    
+    bilog_autotrim(secondary.zone)
```

Comment:
i tend to think we should do the `bilog_autotrim()` and `check_bucket_instance_metadata()` on every zone, not just `rw_zones[0]` and `rw_zones[1]`. this would help make sure we get full test coverage for multi-zonegroups

Reply:
for multi-zonegroups shouldn't we just skip other zonegroups as they can't host any data for those buckets hosted in the bucket's zonegroup?

Reply:
@clwluvw currently, the bucket would still create indexes on other zonegroups. so upon bucket deletion, we should also remove all the metadata associated with it.

Reply:
I don't think so, from here (https://github.com/ceph/ceph/blob/90d0f0306f9a1fe6e78bdda5c29fc5137dcafda6/src/rgw/rgw_op.cc#L3639-L3642) only when bucket's zonegroup is the RGW's zonegroup `createparams.zone_placement` will be assigned and then in `RGWRados::create_bucket` it only creates the indexes when `zone_placement` is not empty: https://github.com/ceph/ceph/blob/90d0f0306f9a1fe6e78bdda5c29fc5137dcafda6/src/rgw/driver/rados/rgw_rados.cc#L2362-L2367

Reply:
Do you have those commits In your branch?
It works for me as i have faced problems like this: https://github.com/ceph/ceph/pull/60227

Reply:
Can we have this commit (https://github.com/ceph/ceph/pull/59911/commits/be9ad44e5a7abb5f9d14ea89a6d46f4655c99c5e) or a check like this (https://github.com/ceph/ceph/blob/be9ad44e5a7abb5f9d14ea89a6d46f4655c99c5e/src/rgw/driver/rados/rgw_bucket.cc#L2820-L2824) backported here so we can make sure those indexes won't be created even through the meta sync?

Reply:
as far as this pr is concerned, the cleanup works perfectly well on multiple zonegroups. so just making the test cases pass is what is needed here. even if the bucket is indexless, we will still have bucket instance left behind to be cleaned up. maybe we can consider indexless strategy separately outside this pr if it is well-motivated.

Reply:
for multi-zonegroup, i could see a design where we let metadata sync (via `RGWBucketInstanceMetadataHandler::remove()`) delete the bucket instance metadata for buckets that belong to other zonegroups, and only defer that deletion until bilog trimming for local buckets. but i'm not sure it's worth treating local/remote buckets differently if the trimming solution can work for both

but who deletes the bucket index shards? i had thought `RGWBucketInstanceMetadataHandler::remove()` handled that, but i don't see a call to `clean_index()` there

Reply:
> but i'm not sure it's worth treating local/remote buckets differently if the trimming solution can work for both

I had a concern about unnecessary calls when it came to the scale of x zonegroups and x deletions. Perhaps hard to measure now but if we can eliminate why not? :)

> but who deletes the bucket index shards? i had thought RGWBucketInstanceMetadataHandler::remove() handled that, but i don't see a call to clean_index() there

We could add the if clause I mentioned earlier to prevent initializing indexes for remote buckets. I originally suggested using it for an indexless approach, but even if we don't take that route now, the check is still worth having. so we won't need to clean_index() at all for remote_buckets.

Reply:
> Perhaps hard to measure now but if we can eliminate why not? :)

yeah.. bilog trimming makes its decisions based on the peer zones' sync status with respect to the local bilog. but when it runs on a remote zonegroup, there is no local bilog so no peers will report any status for it. empty status probably still causes us to handle the cleanup, but i agree that those http requests to fetch sync status aren't useful there

> so we won't need to clean_index() at all for remote buckets.

right, but we do need to call it for local buckets. i _think_ that should be part of the bucket instance metadata's deletion, which itself knows whether its indexless so can do the right thing

Reply:
> right, but we do need to call it for local buckets. i think that should be part of the bucket instance metadata's deletion, which itself knows whether its indexless so can do the right thing

Right, but I got confused why do we need to keep indexes when `is_syncing_bucket_meta()` and not call `CLSRGWIssueBucketIndexClean()`? I guess the status object is not stored in index so what is the point of having keeping it?

Reply:
> right, but we do need to call it for local buckets. i _think_ that should be part of the bucket instance metadata's deletion, which itself knows whether its indexless so can do the right thing

sorry, Shilpa clarified this in the refactoring call. trim is already handling index object deletion via `BucketCleanIndexCollectCR`

Reply:
Just got a general question regarding `BucketCleanIndexCollectCR`, in multi-zonegroup (two zonegroup one in each) scenario in the current main branch implementation how is `BucketCleanIndexCollectCR` going to be called? Isn't that going to be stopped here: https://github.com/ceph/ceph/blob/7e6c5b8573955b5bdc5b6ca74aa5a06233a6f2cf/src/rgw/driver/rados/rgw_trim_bilog.cc#L619-L630

Reply:
`RGWBucketGetSyncPolicyHandlerCR` is just reading the local zone's sync policy and returns a default handler if there isn't one. further down, we query peer zones of that zonegroup for sync status with a `get_all_dests()` and expect a response based on which we decide whether to trim a particular generation or not.
if you are asking about having a single zone in each zonegroup, RGWBucketGetSyncPolicyHandlerCR will still give you a default(empty) policy. so won't error out here. but it will go ahead and simply trim everything since it doesn't have any peers to query. this is something we will have to consider in cross-zonegroup replication.

Reply:
right, I forgot we create one policy handler per zone on startup https://github.com/ceph/ceph/blob/c0dae4614e22751b835928fee65d49b4a015f090/src/rgw/services/svc_zone.cc#L300